triggers/service/authentication.md

```diff
@@ -0,0 +1,95 @@
+# Design for Trigger Service Authentication/Authorization
+
+## Goals
+
+- Be compatible with an OAuth2 authorization code grant
+  https://tools.ietf.org/html/rfc6749#section-4.1
+- Do not require OAuth2 or any other specific
+  authentication/authorization protocol from the IAM. In other words,
+  the communication with the IAM must be pluggable.
+- Do not rely on wildcard access for the trigger service, it should
+  only be able to start triggers on behalf of a party if a user that
+  controls that party has given consent.
+- Support long-running triggers without constant user
+  interaction. Since auth tokens are often short-lived (e.g., expire
+  after 1h), this implies some mechanism for token refresh.
+
+## Design
+
+This involves 3 components:
+
+1. The trigger service provided by DA.
+2. An auth middleware. DA provides an implementation of this for at
+   least the OAuth2 authorization code grant but this is completely
+   pluggable so if the DA-provided middleware does not cover the IAM
+   infrastructure of a client, they can implement their own.
+3. The IAM. This is the entity that signs Ledger API tokens. This is
+   not provided by DA. The Ledger is configured to trust this entity.
+
+### Auth Middleware API
+
+The auth middleware provides a few endpoints (the names don’t matter
+all that much, they just need to be fixed once).
+
+1. /auth The trigger service, will contact this endpoint with a set of
```

Comment:
ok in this sense the auth middleware will trust an upstream identity provider which implies that there is a requirement for the claims as furnished by that provider and the claims that the middleware understands, right?

Reply:
The middleware has to translate between claims (a ledger concept) and OAuth2 claims/custom parameters in the authorize request/…. We assume that at the end we do get a ledger token back from the middleware which in turn gets it from the authorization server.